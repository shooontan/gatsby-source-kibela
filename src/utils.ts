import camelCase from 'camelcase';
import crypto from 'crypto';
import { Actions, NodePluginArgs } from 'gatsby';

type CreateContentNodeArgs = {
  actions: Pick<Actions, 'createNode'>;
  createNodeId: NodePluginArgs['createNodeId'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  content: any;
  type: string;
};

const typePrefix = 'Kibela';
const makeTypeName = (type: string) =>
  camelCase([typePrefix, type], { pascalCase: true });

export const createContentNode = ({
  actions: { createNode },
  createNodeId,
  content,
  type,
}: CreateContentNodeArgs) => {
  const nodeContent = JSON.stringify(content);
  const nodeId = createNodeId(content?.id || nodeContent);
  const nodeContentDigest = crypto
    .createHash('md5')
    .update(nodeContent)
    .digest('hex');

  const node = {
    ...content,
    id: nodeId,
    parent: null,
    children: [],
    internal: {
      type: makeTypeName(type),
      content: nodeContent,
      contentDigest: nodeContentDigest,
    },
  };

  /**
   * To avoid conflicting named id,
   * rename to ${type}Id if content has id property.
   */
  if (content.id) {
    const contentIdName = camelCase([type, 'id']);
    node[contentIdName] = content.id;
  }

  createNode(node);
};
