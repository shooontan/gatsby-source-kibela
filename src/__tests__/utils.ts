import { createContentNode } from '../utils';

const createNode = jest.fn();
const createNodeId = jest.fn();

describe('createContentNode', () => {
  beforeEach(() => {
    createNode.mockClear();
    createNodeId.mockClear();
  });

  test('create node', () => {
    createNodeId.mockReturnValue('nodeId');
    createContentNode({
      actions: {
        createNode,
      },
      createNodeId,
      content: { title: 'hello' },
      type: 'note',
    });
    expect(createNode.mock.calls).toMatchSnapshot();
    expect(createNode.mock.calls[0][0].title).toBe('hello');
  });

  test('content node with id', () => {
    createNodeId.mockReturnValue('nodeId');
    createContentNode({
      actions: {
        createNode,
      },
      createNodeId,
      content: { id: 'uYarjeb7YL', title: 'hello' },
      type: 'note',
    });
    expect(createNode.mock.calls).toMatchSnapshot();
    expect(createNode.mock.calls[0][0].id).not.toBe('uYarjeb7YL');
    expect(createNode.mock.calls[0][0].noteId).toBe('uYarjeb7YL');
  });
});
