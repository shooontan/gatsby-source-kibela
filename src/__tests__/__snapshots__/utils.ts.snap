// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContentNode content node with id 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "id": "nodeId",
      "internal": Object {
        "content": "{\\"id\\":\\"uYarjeb7YL\\",\\"title\\":\\"hello\\"}",
        "contentDigest": "955cc55ca070043c9e9847c96cd651f9",
        "type": "KibelaNote",
      },
      "noteId": "uYarjeb7YL",
      "parent": null,
      "title": "hello",
    },
  ],
]
`;

exports[`createContentNode create node 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "id": "nodeId",
      "internal": Object {
        "content": "{\\"title\\":\\"hello\\"}",
        "contentDigest": "e90d9516ef7a6840faa8bc08e5b5cf0c",
        "type": "KibelaNote",
      },
      "parent": null,
      "title": "hello",
    },
  ],
]
`;
